bigchaindb/models.py

```diff
@@ -0,0 +1,209 @@
+                                      .format(input_txid))
+        else:
+            raise TypeError('`operation` must be either `TRANSFER`, `CREATE` '
+                            'or `GENESIS`')
+
+        if not self.fulfillments_valid(input_conditions):
+            raise InvalidSignature()
+        else:
+            return self
+
+
+class Block(object):
+    def __init__(self, transactions=None, node_pubkey=None, timestamp=None,
+                 voters=None, signature=None):
+        if transactions is not None and not isinstance(transactions, list):
+            raise TypeError('`transactions` must be a list instance or None')
+        elif transactions is None:
+            self.transactions = []
+        else:
+            self.transactions = transactions
+
+        if voters is not None and not isinstance(voters, list):
+            raise TypeError('`voters` must be a list instance or None')
+        elif transactions is None:
+            self.voters = []
+        else:
+            self.voters = voters
+
+        if timestamp is not None:
+            self.timestamp = timestamp
+        else:
+            self.timestamp = gen_timestamp()
+
+        self.node_pubkey = node_pubkey
+        self.signature = signature
+
+    def __eq__(self, other):
+        try:
+            other = other.to_dict()
+        except AttributeError:
+            return False
+        return self.to_dict() == other
+
+    def validate(self, bigchain):
+        """Validate a block.
+
+        Args:
+            bigchain (Bigchain): an instantiated bigchaindb.Bigchain object.
+
+        Returns:
+            block (Block): The block as a `Block` object if it is valid.
+                           Else it raises an appropriate exception describing
+                           the reason of invalidity.
+
+        Raises:
+            OperationError: if a non-federation node signed the block.
+        """
+
+        # First, make sure this node hasn't already voted on this block
+        if bigchain.has_previous_vote(self.id, self.voters):
+            return self
+
+        # Check if the block was created by a federation node
+        possible_voters = (bigchain.nodes_except_me + [bigchain.me])
+        if self.node_pubkey not in possible_voters:
+            raise OperationError('Only federation nodes can create blocks')
+
+        if not self.is_signature_valid():
+            raise InvalidSignature('Block signature invalid')
+
+        # Finally: Tentative assumption that every blockchain will want to
+        # validate all transactions in each block
+        for tx in self.transactions:
+            # NOTE: If a transaction is not valid, `is_valid` will throw an
+            #       an exception and block validation will be canceled.
+            bigchain.validate_transaction(tx)
+
+        return self
+
+    def sign(self, private_key):
+        block_body = self.to_dict()
+        block_serialized = serialize(block_body['block'])
```

Comment:
> Maybe it's worth adding a method that just creates the block dict?

You mean this part? https://github.com/bigchaindb/bigchaindb/pull/641/files/76c796429055d180dcf3caaf2a95e1591e0d743c#diff-9f1ab5b5daf1b3072e2123b6939bfb07R193 ?